File: Exercises/MPI/Examples/solution1.f
      program shifter
      implicit none

      include 'mpif.h'

      integer my_pe_num, errcode, numbertosend, message_received
      integer status(MPI_STATUS_SIZE)

      call MPI_INIT(errcode)

      call MPI_COMM_RANK(MPI_COMM_WORLD, my_pe_num, errcode)

      numbertosend = my_pe_num

      if (my_pe_num.EQ.7) then
         call MPI_Send(numbertosend, 1, MPI_INTEGER, 0, 10,
     +        MPI_COMM_WORLD, errcode)
      else
         call MPI_Send(numbertosend, 1, MPI_INTEGER, my_pe_num+1, 10,
     +        MPI_COMM_WORLD, errcode)
      endif

      call  MPI_Recv(message_received, 1, MPI_INTEGER, MPI_ANY_SOURCE,
     +     10, MPI_COMM_WORLD, status, errcode)

      print *,'PE', my_pe_num, ' received ', message_received, '.'

      call MPI_FINALIZE(errcode)
      end

File: Exercises/MPI/Examples/solution1.c
#include <stdio.h>
#include "mpi.h"

main(int argc, char** argv){

  int my_PE_num, number_to_send, message_received;
  MPI_Status status;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &my_PE_num);

  number_to_send = my_PE_num;

  if (my_PE_num==7)
      MPI_Send( &number_to_send, 1, MPI_INT, 0, 10, MPI_COMM_WORLD);
  else
      MPI_Send( &number_to_send, 1, MPI_INT, my_PE_num+1, 10, MPI_COMM_WORLD);

  MPI_Recv( &message_received, 1, MPI_INT, MPI_ANY_SOURCE, 10, MPI_COMM_WORLD, &status);

  printf("PE %d received %d.\n", my_PE_num, message_received);
  
  MPI_Finalize();

}
